### PR TITLE
DBaaS Platform Catalog design

### PR DESCRIPTION
**Pull request by denisok**
https://jira.percona.com/browse/PMM-11065

https://github.com/Percona-Lab/dbaas-catalog/blob/catalog-design/docs/design/dbaas-catalog-design.md
https://github.com/Percona-Lab/dbaas-catalog/blob/catalog-design/docs/design/Percona-DBaaS-Platform-design.md

### CODE REVIEW

File: docs/design/dbaas-catalog-design.md [hunk 1]
```diff
@@ -0,0 +1,785 @@
...
+-->
+
+#### Story 1
+
+As SRE I would like to add one catalog and be abe to deploy my variant of DBaaS Platform from it.
+
+For example:
+  - install DB Server operator
+  - install monitoring operator
+  - install Database operators that help to manage and operate Databases
```
Comment (gen1us2k) [inline docs/design/dbaas-catalog-design.md:129]: Here are some additional user stories 

As an SRE I should be able to install any operator from the catalog and subscribe to updates to be able to have updated versions of operators in my kubernetes cluster. 

As an SRE I should be able to switch between catalog channels to try the newest versions of database operators to test my upgrade plans to understand the consequences of upgrading and possible downtime on the production environment.

File: docs/design/dbaas-catalog-design.md [hunk 2]
```diff
@@ -0,0 +1,785 @@
...
+
+#### Story 2
+
```
Comment (gen1us2k) [inline docs/design/dbaas-catalog-design.md:132]: As a DBaaS platform developer, I should be able to install the newest version of a database operator and run automated tests against it to improve the overall quality of the operator's interoperability 

As a DBaaS QA, I should be able to install any version of a database operator or a set of database operators in order to test the update/upgrade scenarios and the operator's interoperability.

File: docs/design/dbaas-catalog-design.md [hunk 3]
```diff
@@ -0,0 +1,785 @@
...
+```sh
+mkdir catalog/percona-xtradb-cluster-operator ; cd catalog/percona-xtradb-cluster-operator
+cat << EOF >> veneer.yaml
+Schema: olm.semver
+GenerateMajorChannels: true
+GenerateMinorChannels: false
+Stable:
+  Bundles:
+  - Image: quay.io/operatorhubio/percona-xtradb-cluster-operator:v1.10.0
```
Comment (gen1us2k) [inline docs/design/dbaas-catalog-design.md:240]: I assume that for any operator that we develop there should be the release process that we have right now. An operator's team should create a new semver release tag for a new operator, bumping the version that was released, right? And in that case, we can use Fast (dev), Canditate (on release cut), and stable (Release) versions.

During the active development phase, operator image will be updated automatically using pipelines? Just to run tests against the newest version of an operator. Did I understand correctly the process of developing/testing operators?

File: docs/design/dbaas-catalog-design.md [hunk 4]
```diff
@@ -0,0 +1,785 @@
...
+
+```
+podman build . -f dbaas-catalog.Dockerfile -t ghcr.io/percona-platform/dbaas-catalog:main
+podman push ghcr.io/percona-platform/dbaas-catalog:main
+```
+
+Questions:
+  - could it be links to the veneers in other GH repos instead (submodules)?
+
+### Creating new platform
```
Comment (gen1us2k) [inline docs/design/dbaas-catalog-design.md]: What is a platform here?

**Review by gen1us2k — APPROVED** [review]

File: docs/design/dbaas-catalog-design.md (hunk 4 above)
Comment (denisok) [reply docs/design/dbaas-catalog-design.md]: added explanation.

**Comment by denisok** [comment]
@atymchuk @hors @egegunes @gen1us2k please review this one.

We need to identify good repo place and registry for this one. I wonder if:

- github.com/percona-platform/dbaas-catalog
- docker.io/percona-platform

is a good idea to place this repo and publish images?

File: docs/design/Percona-DBaaS-Platform-design.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  - **Candidate** Channel: `docker.io/percona-platform/operatorA:vX.Y.Z-rcN`
+  - **Stable** Channel: `docker.io/percona-platform/operatorA:vX.Y.Z`
+
+### Owners
+
+Platform owners are responsible and accountable for the platform CI/CD, quality (green actions) and state of the operators (new versions) in the catalog.
+
+Platform owners:
+  - @denisok
+  - @nmarukovich
```
Comment (nmarukovich) [inline docs/design/Percona-DBaaS-Platform-design.md:94]: @denisok could you add one more person to owner list? I could be @cap1984. Do you plan to add pg operator?

File: docs/design/dbaas-catalog-design.md [hunk 6]
```diff
@@ -0,0 +1,563 @@
...
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a SRE I would like to add one catalog and be abe to deploy my variant of DBaaS Platform from it.
+
```
Comment (nmarukovich) [inline docs/design/dbaas-catalog-design.md:125]: abe -> able

File: docs/design/dbaas-catalog-design.md [hunk 7]
```diff
@@ -0,0 +1,563 @@
...
+This Catalog will be based on  [Veneers](https://olm.operatorframework.io/docs/reference/veneers/), specifically on `semver veneer` as an approach to simplifying interacting with FBCs (File-Based Catalogs).
+
+Please see [Semver Veneer specification](https://olm.operatorframework.io/docs/reference/veneers/#semver-veneer).
+
+Catalog should have number of different testing layers to test:
+  - sanity of individual operator veneer/catalog
+  - integration test of new catalog with updated operator
+  - interoperability of the new operator version with other operators that could be affected
+  - e2e test for the platform configurations
+
```
Comment (nmarukovich) [inline docs/design/dbaas-catalog-design.md:198]: Do you have any ideas what exactly do you want to test on different level? Do you plan to mock k8s in integration tests? 
- interoperability of the new operator version with other operators that could be affected
How does this connection look?

Comment (denisok) [reply docs/design/dbaas-catalog-design.md:198]: for example PR could have some basic sanity when more integration tests are run on `main` commit.

As how those test would look like - no idea so far :)

There should be tests, but how they would look like - TBD.

File: docs/design/dbaas-catalog-design.md [hunk 8]
```diff
@@ -0,0 +1,563 @@
...
+As a SRE I would like to add one catalog and be able to deploy my variant of DBaaS Platform from it.
+
+For example:
+  - install DB Server operator
+  - install monitoring operator
+  - install Database operators that help to manage and operate Databases
+
+#### Story 2
+
+As a SRE I should be able to install any operator from the catalog and subscribe to updates to be able to have updated versions of operators in my kubernetes cluster.
```
Comment (maxkondr) [inline docs/design/dbaas-catalog-design.md:133]: should we cover DB cluster update as well? or any product that operator deploys (monitoring, logging, etc)

File: docs/design/dbaas-catalog-design.md [hunk 9]
```diff
@@ -0,0 +1,563 @@
...
+  - sanity of individual operator veneer/catalog
+  - integration test of new catalog with updated operator
+  - interoperability of the new operator version with other operators that could be affected
+  - e2e test for the platform configurations
+
+This repo provides only Catalog, OLM [operator bundles](https://olm.operatorframework.io/docs/tasks/creating-operator-bundle/) and their images are provided and maintained by individual operators.
+
+There could be number of bundle sources for this DBaaS Platform Catalog. Operator owners could:
+  1. create their own bundles and images
+  2. push operator's manifest and metadata to the [Kubernetes Community operators](https://github.com/k8s-operatorhub/community-operators) repo
```
Comment (maxkondr) [inline docs/design/dbaas-catalog-design.md:203]: didn't get this idea.
Are we going to copy operators from K8S community hub into our own? Or how is it going to be used?

File: docs/design/dbaas-catalog-design.md (hunk 8 above)
Comment (denisok) [reply docs/design/dbaas-catalog-design.md:133]: operator does DB Cluster update

**Review by gen1us2k — APPROVED** [review]
LGTM

**Comment by denisok** [comment]
> @atymchuk @hors @egegunes @gen1us2k please review this one.
> 
> We need to identify good repo place and registry for this one. I wonder if:
> 
>     * github.com/percona-platform/dbaas-catalog
> 
>     * docker.io/percona-platform
> 
> 
> is a good idea to place this repo and publish images?

during a meeting we decided to go with `percona` repo and registry. design updated.

**Review by egegunes — APPROVED** [review]

**Pull request merged**